multimodal/vl2l/src/mlperf_inference_multimodal_vl2l/evaluation.py

```diff
@@ -0,0 +1,176 @@
+                intersection_count += 1
+            else:
+                # Stop as soon as a mismatch is found (enforces hierarchical match)
+                break
+
+        pred_length = len(predicted_categories)
+        true_length = len(true_categories)
+
+        return intersection_count, pred_length, true_length
+
+    def calculate_hierarchical_metrics(self, data: list) -> float:
+        """Calculates the aggregate hP, hR, and hF scores for a list of samples.
+
+        Args:
+            data: A list of tuples, where each tuple is
+                (predicted_path_str, true_path_str).
+
+        Returns:
+            F1 score
+        """
+        total_intersection = 0
+        total_predicted_length = 0
+        total_true_length = 0
+
+        # 1. Aggregate the components across all samples
+        for pred_path, true_path in data:
+            intersection, pred_len, true_len = \
+                self.get_hierarchical_components(pred_path, true_path)
+
+            total_intersection += intersection
+            total_predicted_length += pred_len
+            total_true_length += true_len
+
+        # 2. Calculate hP and hR
+        hp = total_intersection / total_predicted_length \
+            if total_predicted_length > 0 else 0.0
+        hr = total_intersection / total_true_length \
+            if total_true_length > 0 else 0.0
+
+        return 0.0 if hp + hr == 0 else 2 * (hp * hr) / (hp + hr)
+
+    def calculate_exact_match(self, generated_text: str, original_text: str) -> float:
+        """Calculates binary Exact Match (EM) score.
+
+        We clean the text (lowercase, strip whitespace) for a fairer comparison.
+
+        Args:
+            generated_text: Output from the VLM.
+            original_text: Ground truth information from the dataset.
+
+        Returns:
+            1 if the values match or 0 otherwise
+        """
+        gen = generated_text.strip().lower()
+        orig = original_text.strip().lower()
+
+        return 1.0 if gen == orig else 0.0
+
+
+    def run_evaluation(self) -> None:
+        """Main function to run the evaluation."""
+        with Path.open(self.filename) as f:
+            model_output = json.load(f)
+
+        original_data = load_dataset(
+            self.dataset.repo_id,
+            token=self.dataset.token,
+        )["train"]
+
+        category_dataset_pred_src = []
+        is_secondhand_pred = []
+        is_secondhand_src = []
+        for elem in model_output:
+            byte_data = bytes.fromhex(elem["data"])
+            idx = elem["qsl_idx"]
+            pred_text_decode = byte_data.decode("utf-8")
+            """
+            Model response is similar to:
+            ```json
+             {
+             .....
+             }
+            ```
+            Need to find open and close brackets
+            """
+            start_index = pred_text_decode.find("{")
+            end_index = pred_text_decode.rfind("}") + 1
```

Comment:
Related to [this comment](https://github.com/mlcommons/inference/pull/2393/files#r2543623303). We should just ask the model to give a valid JSON object. If the model doesn't follow the instruction, it's the model's fault.